### DISCUSSION TITLE
Q: CI pipelines with multiple stages/jobs

### DISCUSSION THREAD

**Discussion by gaui** [dagger/dagger#2081] [General]
This project seems promising, but regarding https://docs.dagger.io/1201/ci-environment and CI pipelines

With multiple stages and jobs, wouldn't you always have to execute each dagger action independently to be able to use parallelism and autoscaling of pipeline agents/workers? So in fact you are always tied to writing a pipeline file (for those platforms that support Pipelines as Code) and call dagger actions from that?

**Comment by gerhard** [comment]
You will need to write a pipeline file specific to the CI environment in order to run `dagger do <action>` commands. Dagger uses BuildKit to run these actions. By default, BuildKit gets created as a container in the local Docker instance. If you try running this locally, you will see the `dagger-buildkitd` container running. All actions will be represented as a DAG, and many will be cached. Subsequent runs will be faster.

Now imagine that you have a cluster of BuildKit instances, perhaps running in Kubernetes, and there is a hash ring which determines task placement. Remember, Dagger doesn't dedup tasks - BuildKit does. Running the same task with the same inputs from multiple dagger instances results in a single run, and all others picking it up from the BuildKit cache. This is related: https://www.youtube.com/watch?v=wTENRhYt3mw

The simplest option of scaling this today - something that you can do right now - is to use an external Docker Engine. This is one use case that does just that: https://github.com/thechangelog/changelog.com/pull/395. Notice the `DOCKER_HOST` which uses a Tailscale tunnel to keep everything secure. It works really well in practice, and we are working on the next step: https://github.com/dagger/dagger/pull/1839

There is also mounts to talk about, but let's leave that for another time 😉

cc @aluzzardi & @shykes
